File: android/app/src/main/kotlin/com/agent/aios/domain/agent/ConfirmationGate.kt
package com.agent.aios.domain.agent

import android.util.Log
import com.agent.aios.domain.model.AgentStep
import com.agent.aios.domain.model.ToolRisk
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class ConfirmationGate {
    private val TAG = "AIOS-Confirm"
    private var latch: CountDownLatch? = null

    @Volatile
    private var approved = false

    fun requestConfirmation(
        risk: ToolRisk,
        toolName: String,
        args: String,
        onStep: (AgentStep) -> Unit,
    ): Boolean {
        latch = CountDownLatch(1)
        approved = false

        onStep(
            AgentStep(
                type = "confirmation_required",
                content = "Requires confirmation: $toolName",
                toolName = toolName,
                toolArgs = args,
                riskLevel = risk.name,
            ),
        )

        return try {
            latch?.await(60, TimeUnit.SECONDS) ?: false
            approved
        } catch (e: Exception) {
            Log.w(TAG, "Confirmation interrupted: ${e.message}")
            false
        } finally {
            latch = null
        }
    }

    fun resolve(approved: Boolean) {
        this.approved = approved
        latch?.countDown()
    }

    fun cancel() {
        approved = false
        latch?.countDown()
    }
}

File: lib/domain/agent/confirmation_gate.dart
import 'package:aios/domain/agent/gate_completer.dart';
import 'package:aios/domain/entities/agent_models.dart';

class ConfirmationGate extends GateCompleter {
  static const _tag = 'AIOS-ConfirmationGate';

  Future<bool> requestConfirmation(
    ToolRisk risk,
    String toolName,
    String args,
    void Function(AgentStep) onStep,
  ) async {
    createCompleter();

    onStep(
      AgentStep(
        'confirmation_required',
        'Requires confirmation: $toolName',
        toolName: toolName,
        toolArgs: args,
        riskLevel: risk.name,
      ),
    );

    return waitForResolution(
      timeout: const Duration(seconds: 60),
      tag: _tag,
      label: 'confirmation for $toolName',
    );
  }
}
